File: Backend/Service/UploadImageService.cs
namespace Backend.Service;

public class UploadImageService
{
    public async Task<string> UploadImage(IFormFile file)
    {
        List<string> validExtensions = new List<string> { ".jpg", ".jpeg", ".png", ".gif" };
        
        string extension = Path.GetExtension(file.FileName).ToLower();
        if (!validExtensions.Contains(extension))
        {
            throw new Exception("Invalid file extension");
        }
        
        if (!file.ContentType.StartsWith("image/"))
        {
            throw new Exception("Invalid file content type");
        }
        
        if (file.Length > 5 * 1024 * 1024) // 5 MB limit
        {
            throw new Exception("File is too large");
        }
        
        string fileName = Guid.NewGuid() + extension;
        string filePath = Path.Combine(Directory.GetCurrentDirectory(), "wwwroot", "Images");
        if (!Directory.Exists(filePath))
        {
            Directory.CreateDirectory(filePath);
        }
            
        string fullPath = Path.Combine(filePath, fileName);
        await using (var stream = new FileStream(fullPath, FileMode.Create))
        {
            await file.CopyToAsync(stream);
        }
            
        return "/Images/" + fileName;
    }
}

File: PKM-Backend-JAVA/src/main/java/com/example/service/UploadImageService.java
package com.example.service;

import com.example.exception.BusinessException;
import com.example.exception.ValidationException;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;

import java.io.InputStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.StandardCopyOption;
import java.util.UUID;

@Service
public class UploadImageService {
    public String uploadImage(MultipartFile file) {
        String[] validExtensions = new String[]{"png", "jpg", "jpeg", "gif", "bmp"};

        if (file == null || file.isEmpty()) {
            throw new ValidationException("Image file is required");
        }

        if (!file.getContentType().startsWith("image/")) {
            throw new ValidationException("Invalid image content type");
        }

        if (file.getSize() > 5 * 1024 * 1024) {
            throw new ValidationException("Image must be under 5MB");
        }

        String originalFilename = file.getOriginalFilename();

        if (originalFilename == null || !originalFilename.contains(".")) {
            throw new ValidationException("Invalid image file name");
        }

        String suffix = originalFilename.substring(originalFilename.lastIndexOf(".")+1).toLowerCase();

        boolean isValid = false;

        for(String validExtension:validExtensions){
            if(validExtension.equals(suffix)){
                isValid = true;
                break;
            }
        }

        if(!isValid){
            throw new ValidationException("Invalid image extension");
        }

        String fileName = UUID.randomUUID().toString()+"."+suffix;
    try{
        Path directory = Paths.get(System.getProperty("user.dir"), "uploads", "images");

        if(!Files.exists(directory))
            Files.createDirectories(directory);

        Path fullPath = directory.resolve(fileName);
        try(InputStream inputStream = file.getInputStream()){
            Files.copy(inputStream,fullPath, StandardCopyOption.REPLACE_EXISTING);
        }

        return "/uploads/images/" + fileName;
    }
    catch (Exception ex) {
        throw new BusinessException("Failed to upload image");
    }
    }
}
